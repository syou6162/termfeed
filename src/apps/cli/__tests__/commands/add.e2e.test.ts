import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import {
  createTestContext,
  runCommand,
  setupRSSCrawlerMock,
  createMockRSSData,
  type TestContext,
} from '@/test-helpers/index.js';

describe('add command E2E', () => {
  let context: TestContext;
  let rssMock: ReturnType<typeof setupRSSCrawlerMock>;

  beforeEach(() => {
    context = createTestContext();
    rssMock = setupRSSCrawlerMock();
  });

  afterEach(() => {
    context.cleanup();
    vi.restoreAllMocks();
  });

  it('should add RSS feed successfully', async () => {
    // Arrange
    const feedUrl = 'https://example.com/feed.rss';
    const mockData = createMockRSSData({
      title: 'Test Feed',
      description: 'Test Description',
      feedUrl,
    });
    rssMock.mockFeedResponse(feedUrl, mockData);

    // Act
    const output = await runCommand(['add', feedUrl], {
      dbPath: context.dbPath,
    });

    // Assert - 終了コード
    expect(output.exitCode).toBeUndefined(); // 正常終了

    // Assert - 出力のスナップショット
    expect(output.stdout).toMatchSnapshot('add-success-output');

    // Assert - データベースの状態（コンテキストのモデルを使って検証）
    const feeds = context.feedModel.findAll();
    expect(feeds).toHaveLength(1);
    expect(feeds[0]).toMatchObject({
      url: feedUrl,
      title: 'Test Feed',
      description: 'Test Description',
    });

    const articles = context.articleModel.findAll();
    expect(articles).toHaveLength(2); // モックデータのデフォルト
  });

  it('should handle network errors gracefully', async () => {
    // Arrange
    const feedUrl = 'https://example.com/error.rss';
    rssMock.mockError(new Error('Network error'));

    // Act
    const output = await runCommand(['add', feedUrl], {
      dbPath: context.dbPath,
    });

    // Assert
    expect(output.exitCode).toBe(1);
    expect(output.stderr).toContain('Error adding feed');

    // データベースに何も追加されていないことを確認
    const feeds = context.feedModel.findAll();
    expect(feeds).toHaveLength(0);
  });

  it('should handle duplicate feed error', async () => {
    // Arrange
    const feedUrl = 'https://example.com/feed.rss';
    const mockData = createMockRSSData({
      title: 'Test Feed',
      description: 'Test Description',
      feedUrl,
    });
    rssMock.mockFeedResponse(feedUrl, mockData);

    // 最初のフィードを追加
    await runCommand(['add', feedUrl], {
      dbPath: context.dbPath,
    });

    // Act - 同じフィードを再度追加
    const output = await runCommand(['add', feedUrl], {
      dbPath: context.dbPath,
    });

    // Assert
    expect(output.exitCode).toBe(1);
    expect(output.stderr).toContain('Error adding feed');

    // フィードは1つのみ
    const feeds = context.feedModel.findAll();
    expect(feeds).toHaveLength(1);
  });

  it('should handle invalid URL gracefully', async () => {
    // Act
    const output = await runCommand(['add', 'not-a-valid-url'], {
      dbPath: context.dbPath,
    });

    // Assert
    expect(output.exitCode).toBe(1);
    expect(output.stderr).toContain('Error adding feed');

    // データベースに何も追加されていないことを確認
    const feeds = context.feedModel.findAll();
    expect(feeds).toHaveLength(0);
  });
});
