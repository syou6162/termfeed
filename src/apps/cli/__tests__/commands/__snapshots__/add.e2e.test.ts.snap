// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`add command E2E > should add RSS feed successfully > add-success-output 1`] = `
"Database migration completed successfully
Adding feed: https://example.com/feed.rss
Feed added successfully!
  ID: 1
  Title: Test Feed
  Description: Test Description
  Articles added: 2
"
`;

exports[`add command E2E > should handle duplicate feed error > duplicate-feed-error-output 1`] = `
"Error adding feed: Feed already exists: https://example.com/feed.rss
"
`;

exports[`add command E2E > should handle invalid URL gracefully > invalid-url-error-output 1`] = `
"Error adding feed: Failed to fetch feed: Failed to parse RSS feed: Invalid URL
"
`;

exports[`add command E2E > should handle network errors gracefully > network-error-output 1`] = `
"Error adding feed: Failed to fetch feed: Network error
"
`;
